File: Java-Spring/src/main/java/org/codecop/lcdnumbers/DigitScaler.java
package org.codecop.lcdnumbers;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Scales x and y of digit lines.
 */
public class DigitScaler {

    private final ScalingRepeater repeater;

    public DigitScaler(ScalingRepeater repeater) {
        this.repeater = repeater;
    }

    public List<Digit> scale(List<Digit> digits, Scaling scaling) {
        return digits.stream(). //
                map(d -> scale(d, scaling)). //
                collect(Collectors.toList());
    }

    private Digit scale(Digit digit, Scaling scaling) {
        if (scaling.none()) {
            return digit;
        }

        List<Line> scaled = new ArrayList<>();

        boolean oddLine = true;
        for (Line line : digit.lines()) {
            Line scaledLine = line.scaleHorizontal(repeater, scaling);
            if (oddLine) {
                scaled.add(scaledLine);
            } else {
                scaled.addAll(repeater.repeat(scaledLine, scaling));
            }
            oddLine = !oddLine;
        }

        return digit.scale(scaled);
    }
}

File: Csharp-NAutowired/LcdNumbers/DigitScaler.cs
using System.Collections.Generic;
using System.Linq;
using NAutowired.Core.Attributes;

namespace Org.Codecop.Lcdnumbers
{
    /// <summary>Scales x and y of digit lines.</summary>
    [Service]
    public class DigitScaler : IDigitScaler
    {
        [Autowired]
        private readonly IScalingRepeater repeater;

        public virtual IList<Digit> Scale(IList<Digit> digits, Scaling scaling)
        {
            return digits.Select(d => Scale(d, scaling)).ToList();
        }

        private Digit Scale(Digit digit, Scaling scaling)
        {
            if (scaling.HasNone())
            {
                return digit;
            }

            List<Line> scaled = new List<Line>();

            bool oddLine = true;
            foreach (Line line in digit.Lines())
            {
                Line scaledLine = line.ScaleHorizontal(repeater, scaling);
                if (oddLine)
                {
                    scaled.Add(scaledLine);
                }
                else
                {
                    scaled.AddRange(repeater.Repeat(scaledLine, scaling));
                }
                oddLine = !oddLine;
            }

            return digit.Scale(scaled);
        }
    }
}
